### DISCUSSION TITLE
POST /ollama/api/chat HTTP/1.1 401 Unauthorized

### DISCUSSION THREAD

**Discussion by MartinCanto** [open-webui/open-webui#6856] [Issues]
I am encountering an issue regarding header-based authentication in Open Web UI with a fresh Docker installation. While I'm able to log in, pull and delete models, and perform other actions (e.g., create a new user and view chat history), I get a "Not Authenticated" error whenever I attempt to start a chat. However, the logs suggest that the authentication issue only affects the POST /ollama/api/chat endpoint, which returns a 401 Unauthorized .

When I remove authentication entirely, I am able to chat successfully.

Summary:

Docker-based Open Web UI install
Header-based authentication setup (works for most actions except chat)
Able to log in, pull, delete models, and view settings
Chat initiation fails with a "Not Authenticated" error in the UI
Container logs show 401 Unauthorized related to chatting
Container Logs :

INFO:     10.0.0.2:0 - "GET /api/v1/auths/ HTTP/1.1" 401 Unauthorized
INFO:     10.0.0.2:0 - "GET /api/models HTTP/1.1" 200 OK
INFO:     10.0.0.2:0 - "GET /ollama/api/version HTTP/1.1" 200 OK
INFO:     10.0.0.2:0 - "GET /api/v1/users/user/settings HTTP/1.1" 200 OK
INFO:     10.0.0.2:0 - "GET /api/v1/users/user/settings HTTP/1.1" 200 OK
INFO:     10.0.0.2:0 - "GET /static/favicon.png HTTP/1.1" 304 Not Modified
INFO:     10.0.0.2:0 - "POST /api/v1/chats/new HTTP/1.1" 200 OK
INFO:     10.0.0.2:0 - "GET /api/v1/chats/?page=1 HTTP/1.1" 200 OK
INFO:     10.0.0.2:0 - "POST /ollama/api/chat HTTP/1.1" **401 Unauthorized**

As shown in the logs, I'm seeing one main 401 Unauthorized error that's related specifically to the 

"POST /ollama/api/chat" call, preventing me from successfully starting chats.

The issues seem to be isolated to chat functionality when using header-based authentication.

Could you advise on whether there is a misconfiguration regarding authentication headers for the /ollama/api/chat endpoint? 
Is there anything specific about this endpoint, or any configuration differences I should look into?

Thanks in advance for your help!

**Comment by MartinCanto** [comment]
If I disable Authentik forward auth to the proxy provider, then I am able to chat. So, it seems like something with Authentik maybe is either changing or removing something in the backend request to the /ollama/api/chat.   Even if i dont use headers for authentication if i use an external authorization provider it gives me that 401 error, as soon as I turn off the external authentication it works.

**Comment by derek-assurity** [comment]
Seeing the same error when using SSO Trusted headers. I'm using 
```
      WEBUI_AUTH_TRUSTED_EMAIL_HEADER: 'x-user-email'
      WEBUI_AUTH_TRUSTED_NAME_HEADER: 'x-user-name'
```
Login ok, but get the following error trying to post to Ollama:

` { detail: "Not authenticated" }`
Only seeing this after moving to `v0.3.35`

**Reply by derek-assurity** [reply]
Edit ^ this is also occurring in earlier versions.

**Comment by derek-assurity** [comment]
The problem is being caused by this code -> https://github.com/open-webui/open-webui/blob/96c865404d36637eafadb6d2dd2365c85d452648/backend/open_webui/main.py#L768

It only checks for an `authorization` header, and does not check for the `token` in the cookie. This will break for anyone working with an OAuth proxy in front of Open WebUI. The only workaround I've figured so far is to copy the `token` out of the cookie and add it the request header.

**Comment by derek-assurity** [comment]
Hi @tjbck  , is there any reason that the `token` check in the cookie not used? I can raise a PR to fix otherwise.
Thanks

**Reply by MartinCanto** [reply]
If I understand correctly I will have to copy the token out of the open web ui request cookie and move it to the authorization header?

**Reply by derek-assurity** [reply]
Yes, if you're using something to handle the OAuth side, then you will.

**Reply by bnodnarb** [reply]
Hi @MartinCanto - Can you share the steps you're taking here to copy the token out of the request cookie? Seems like a short term work around.

**Comment by bnodnarb** [comment]
May be worth noting that I'm having the same issue here. I have a reverse proxy (NGINX) in front of my Ollama service on a remote server which checks for Authorization header.

I am able to verify the connection, add models, etc... but when I try to chat I get 401 Unauthorized like the OP does.

Thanks!

**Comment by bnodnarb** [comment]
Hi @MartinCanto and @derek-assurity - I was able to identify the source of this issue:

 The main.py file for Ollama did not fully implement the authorization header for all Ollama endpoints. Namely, the `/api/chat` and `/api/pull` endpoints were not sending out the headers (see example below).
 
 ```
 Successful Requests:

66.8.245.246 - - [24/Nov/2024:21:20:40 +0000] "GET /api/version HTTP/1.1" 200 20 "-" "Python/3.11 aiohttp/3.10.8" Authorization: "Bearer [TOKEN]"
66.8.245.246 - - [24/Nov/2024:21:20:54 +0000] "GET /api/tags HTTP/1.1" 200 660 "-" "python-requests/2.32.3" Authorization: "Bearer [TOKEN]"
66.8.245.246 - - [24/Nov/2024:21:21:07 +0000] "DELETE /api/delete HTTP/1.1" 200 0 "-" "python-requests/2.32.3" Authorization: "Bearer [TOKEN]"
66.8.245.246 - - [24/Nov/2024:21:21:08 +0000] "GET /api/tags HTTP/1.1" 200 338 "-" "Python/3.11 aiohttp/3.10.8" Authorization: "Bearer [TOKEN]"

Failed Requests (bearer not being sent):

66.8.245.246 - - [24/Nov/2024:21:21:32 +0000] "POST /api/pull HTTP/1.1" 401 188 "-" "Python/3.11 aiohttp/3.10.8" Authorization: "-"
66.8.245.246 - - [24/Nov/2024:21:22:08 +0000] "POST /api/chat HTTP/1.1" 401 188 "-" "Python/3.11 aiohttp/3.10.8" Authorization: "-"
```

I have resolved the issue in my forked repository and will be submitting a pull request to the dev branch momentarily. I'll post a link to the pull request once it's been submitted.

Thanks!

**Reply by bnodnarb** [reply]
PR #7326 has been opened to fix. Thanks!

**Reply by MartinCanto** [reply]
Sorry about the late reply, been on vacation, thank you, can't wait for this PR.

**Reply by tomerh2001** [reply]
@bnodnarb Any updates? Still having this issue, can't use the service
